staticaddr/loopin/manager.go

```diff
@@ -0,0 +1,345 @@
+	// enabling it to create and manage loop-ins.
+	DepositManager DepositManager
+
+	// LndClient is used to add invoices and select hop hints.
+	LndClient lndclient.LightningClient
+
+	// InvoicesClient is used to subscribe to invoice settlements and
+	// cancel invoices.
+	InvoicesClient lndclient.InvoicesClient
+
+	// SwapClient is used to get loop in quotes.
+	QuoteGetter QuoteGetter
+
+	// NodePubKey is used to get a loo-in quote.
+	NodePubkey route.Vertex
+
+	// WalletKit is the wallet client that is used to derive new keys from
+	// lnd's wallet.
+	WalletKit lndclient.WalletKitClient
+
+	// ChainParams is the chain configuration(mainnet, testnet...) this
+	// manager uses.
+	ChainParams *chaincfg.Params
+
+	// Chain is the chain notifier that is used to listen for new
+	// blocks.
+	ChainNotifier lndclient.ChainNotifierClient
+
+	// Signer is the signer client that is used to sign transactions.
+	Signer lndclient.SignerClient
+
+	// Store is the database store that is used to store static address
+	// loop-in related records.
+	Store StaticAddressLoopInStore
+
+	// ValidateLoopInContract validates the contract parameters against our
+	// request.
+	ValidateLoopInContract ValidateLoopInContract
+}
+
+// Manager manages the address state machines.
+type Manager struct {
+	cfg *Config
+
+	runCtx context.Context
+
+	sync.Mutex
+
+	// initChan signals the daemon that the address manager has completed
+	// its initialization.
+	initChan chan struct{}
+
+	// currentHeight stores the currently best known block height.
+	currentHeight uint32
+}
+
+// NewManager creates a new deposit withdrawal manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:      cfg,
+		initChan: make(chan struct{}),
+	}
+}
+
+// Run runs the static address loop-in manager.
+func (m *Manager) Run(ctx context.Context, currentHeight uint32) error {
+	m.runCtx = ctx
+
+	m.Lock()
+	m.currentHeight = currentHeight
+	m.Unlock()
+
+	registerBlockNtfn := m.cfg.ChainNotifier.RegisterBlockEpochNtfn
+	newBlockChan, newBlockErrChan, err := registerBlockNtfn(m.runCtx)
+	if err != nil {
+		return err
+	}
+
+	// Upon start of the loop-in manager we reinstate all previous loop-ins
+	// that are not yet completed.
+	err = m.recover()
+	if err != nil {
+		return err
+	}
+
+	// Communicate to the caller that the address manager has completed its
+	// initialization.
+	close(m.initChan)
+
+	for {
+		select {
+		case height := <-newBlockChan:
+			m.Lock()
+			m.currentHeight = uint32(height)
+			m.Unlock()
+
+		case err = <-newBlockErrChan:
+			return err
+
+		case <-m.runCtx.Done():
+			return m.runCtx.Err()
+		}
+	}
+}
+
+// recover stars a loop-in state machine for each non-final loop-in to pick up
+// work where it was left off before the restart.
+func (m *Manager) recover() error {
+	log.Infof("Recovering static address loop-ins...")
+
+	// Recover loop-ins.
+	loopIns, err := m.cfg.Store.AllLoopIns(m.runCtx)
```

Comment:
Would be nice to just select pending ones.

Reply:
Done that, please refer to `GetStaticAddressLoopInSwapsByStates` from `static_address_loopin.sql`.